src/mono/msbuild/apple/build/AppleBuild.targets

```diff
@@ -110,6 +88,54 @@
     </ItemGroup>
   </Target>
 
+  <Target Name="_ScanAssembliesDecideLightweightMarshaler" DependsOnTargets="_AppleResolveReferences">
+    <PropertyGroup Condition="'$(RuntimeComponents)'=='*'">
+      <ForceAllComponents>true</ForceAllComponents>
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(ForceAllComponents)'!='true'">
+      <RuntimeComponentList Include="$(RuntimeComponents)" />
+      <AssembliesToScan Include="@(_AssembliesToBundleInternal)" />
+    </ItemGroup>
+
+    <MarshalingPInvokeScanner Assemblies ="@(AssembliesToScan)" Condition="'$(ForceAllComponents)'!='true'">
+      <Output TaskParameter="IncompatibleAssemblies" ItemName="MonoLightweightMarshallerIncompatibleAssemblies" />
+    </MarshalingPInvokeScanner>
+
+    <ItemGroup Condition="'@(MonoLightweightMarshallerIncompatibleAssemblies->Count())' > 0">
+      <RuntimeComponentList Include="marshal-ilgen" KeepDuplicates="false"/>
+    </ItemGroup>
+    <ItemGroup Condition="'@(MonoLightweightMarshallerIncompatibleAssemblies->Count())' == 0">
+      <RuntimeComponentList Remove="marshal-ilgen" />
+    </ItemGroup>
+
+    <PropertyGroup Condition="'$(ForceAllComponents)'!='true'">
+      <RuntimeComponents>@(RuntimeComponentList)</RuntimeComponents>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="_ProcessRuntimeComponents" DependsOnTargets="_ScanAssembliesDecideLightweightMarshaler" BeforeTargets="_BuildNativeLibrary">
+    <ItemGroup Condition="'$(_IsLibraryMode)' == 'true'">
+      <!-- add all non stub libs first -->
+      <!-- if runtimecompoents is empty, exclude -static.a and include -stub-static.a instead -->
+      <!-- if runtimecomponents is *, we're ok because all -static.a is included -->
+      <!-- if runtimecomponents is a list, add to items and only pull in -static.a -->
+
+      <_UsedComponents 
+        Condition="'$(RuntimeComponents)' != '' and '$(RuntimeComponents)' != '*'"
+        Include="$(RuntimeComponents)" />
+
+      <_RuntimeLibraries 
+        Include="$(AppleBuildDir)\*-stub-static.a" />
+      <_RuntimeLibraries 
+        Include="$(AppleBuildDir)\*.a"
+        Exclude="$(AppleBuildDir)\*-static.a" />
+
+      <_RuntimeLibraries Remove="$(AppleBuildDir)\libmono-component-%(_UsedComponents.Identity)-stub-static.a" />
+      <_RuntimeLibraries Include="$(AppleBuildDir)\libmono-component-%(_UsedComponents.Identity)-static.a" />
```

Comment:
nit: why does the first set of modifications include `*-stub-static.a` but the second one removed `libmono-component-*-stub-static.a` - shouldn't they both use the same kind of pattern (namely: `libmono-component-*-stub-static.a` (and the same for the non-stub components).

I realize this isn't part of your PR - you're just moving it. but it seems like this could be tightened up.

Reply:
What I like about your version is that now we explicitly include everything that doesn't look like a component, and then independently do the component logic.  So in principle we can do those steps separately in the future if there's a need.  It's a bit less coupled now.